File: day04/c/part2.c
#include "helper.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

#define SIZE 10

// const char INPUT[] = "112233-112233";    // test 1
// const char INPUT[] = "123444-123444";    // test 2
// const char INPUT[] = "111122-111122";    // test 3
const char INPUT[] = "136760-595730";    // input


void reverse(int *lst)
{
    int i = 0;
    int j = arrlen(lst) - 1;
    int tmp;
    while (i < j) {
        tmp = lst[i];
        lst[i] = lst[j];
        lst[j] = tmp;
        ++i;
        --j;
    }
}

int * explode(const int number)
{
    int n = number;
    int *digits = NULL;

    while (n > 0) {
        arrput(digits, n % 10);
        n = n / 10;
    }

    reverse(digits);
    return digits;
}

void debug(int *lst)
{
    for (int i = 0; i < arrlen(lst); ++i) {
        printf("%d ", lst[i]);
    }
    puts("");
}

bool is_password(int *digits)
{
    bool ascending = true;

    for (int i = 0; i <= arrlen(digits) - 2; ++i)
    {
        int a = digits[i];
        int b = digits[i + 1];
        if (a > b) {
            ascending = false;
        }
    }
    if (!ascending) {
        return false;
    }

    int counter[SIZE] = { 0 };    // init with 0s
    for (int i = 0; i < arrlen(digits); ++i) {
        const int digit = digits[i];
        counter[digit] += 1;
    }
    bool has_double = false;
    for (int i = 0; i < SIZE; ++i) {
        if (counter[i] == 2) {
            has_double = true;
            break;
        }
    }

    return has_double && ascending;
}

int process(int lo, int hi)
{
    int cnt = 0;
    for (int n = lo; n <= hi; ++n) {
        int *digits = explode(n);
        if (is_password(digits)) {
            ++cnt;
        }
    }

    return cnt;
}

int main()
{
    char* *parts = extract_strings_from_string(strdup(INPUT), "-");
    const int lo = atoi(parts[0]);
    const int hi = atoi(parts[1]);
    // printf("%d %d\n", lo, hi);
    const int result = process(lo, hi);
    printf("%d\n", result);

    return 0;
}

File: day04/rust/src/bin/part2.rs
//const INPUT: &str = "112233-112233";    // test 1
//const INPUT: &str = "123444-123444";    // test 2
//const INPUT: &str = "111122-111122";    // test 3
const INPUT: &str = "136760-595730";    // input


fn explode(number: i32) -> Vec<i32>
{
    let mut n = number;
    let mut digits = Vec::new();
    while n > 0 {
        digits.push(n % 10);
        n = n / 10;
    }

    digits.reverse();
    digits
}

fn is_password(digits: &Vec<i32>) -> bool
{
    let mut ascending = true;

    for i in 0 ..= digits.len()-2 {
        let a = digits[i];
        let b = digits[i+1];
        if a > b {
            ascending = false;
        }
    }
    if !ascending {
        return false;
    }

    let mut counter = [0; 10];
    for digit in digits {
        counter[*digit as usize] += 1;
    }
    let has_double = counter.iter().any(|x| *x == 2);

    has_double && ascending
}

fn process(lo: i32, hi: i32) -> i32
{
    let mut cnt = 0;
    for n in lo ..= hi {
        let digits = explode(n);
        if is_password(&digits) {
            cnt += 1;
        }
    }

    cnt
}

fn main()
{
    let parts: Vec<&str> = INPUT.split('-').collect();
    let lo: i32 = parts[0].parse().unwrap();
    let hi: i32 = parts[1].parse().unwrap();
//    println!("{:?}", parts);
    let result = process(lo, hi);
    println!("{}", result);
}
